File: src/main/scala/demos/fp/order/nohiding/Order.scala
package demos.fp.order.nohiding

case class Order(lines: List[OrderLine] = List(), orderId: String):
    def add(productId: String, amount: Int): Order =
        val existingOrderLineAndIndex =
            this.lines.zipWithIndex.filter(_._1.productId == productId).headOption
        val updatedLines = existingOrderLineAndIndex match
            case None =>
                val newLine = OrderLine(productId, amount)
                this.lines.appended(newLine)
            case Some((existingOrderLine, index)) =>
                val updatedLine = existingOrderLine.add(amount)
                this.lines.updated(index, updatedLine)
        this.copy(lines = updatedLines)

File: src/main/java/demos/oop/order/nohiding/Order.java
package demos.oop.order.nohiding;

import java.util.List;
import java.util.ArrayList;

class Order {
    private List<OrderLine> lines;
    private String orderId;

    public Order(String orderId) {
        this.orderId = orderId;
        this.lines = new ArrayList<OrderLine>();
    }

    public void add(String productId, int amount) {
        OrderLine existingOrderLine = this.findExistingOrderLine(productId);
        if (existingOrderLine == null) {
            OrderLine newOrderLine = new OrderLine(productId, amount);
            newOrderLine.validate();
            this.lines.add(newOrderLine);
        } else {
            existingOrderLine.add(amount);
        }
    }

    private OrderLine findExistingOrderLine(String productId) {
        for (OrderLine orderLine : this.lines) {
            if (orderLine.getProductId().equals(productId)) {
                return orderLine;
            }
        }
        return null;
    }

    public List<OrderLine> getLines() {
        return lines;
    }

    public void setLines(List<OrderLine> lines) {
        this.lines = lines;
    }

    public String getOrderId() {
        return orderId;
    }

    public void setOrderId(String orderId) {
        this.orderId = orderId;
    }

}
